File: app/src/main/java/com/synapse/social/studioasinc/widget/ZoomImageViewLib/gestures/ScaleAndMoveDetector.java
package com.synapse.social.studioasinc.widget.ZoomImageViewLib.gestures;

import android.content.Context;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;

public class ScaleAndMoveDetector implements ScaleGestureDetector.OnScaleGestureListener {
    private float LastX = 0.0f;
    private float LastY = 0.0f;
    private float mCurrentScale = 1.0f;
    private final ScaleGestureDetector mDetector;
    private OnScaleAndMoveGestureListener mDetectorListener;
    private float moveDisFromX = 0.0f;
    private float moveDisFromY = 0.0f;
    private float scaleStart = 1.0f;

    public ScaleAndMoveDetector(Context context, OnScaleAndMoveGestureListener detectorListener) {
        this.mDetector = new ScaleGestureDetector(context, this);
        this.mDetectorListener = detectorListener;
    }

    public boolean onScale(ScaleGestureDetector detector) {
        float tempx = detector.getFocusX();
        float tempy = detector.getFocusY();
        this.moveDisFromX = tempx - this.LastX;
        this.moveDisFromY = tempy - this.LastY;
        if (this.mDetectorListener == null) {
            return false;
        }
        this.mDetectorListener.onScaleAndMove(detector, getCurrentScale(detector.getScaleFactor()), this.moveDisFromX, this.moveDisFromY);
        return false;
    }

    public boolean onScaleBegin(ScaleGestureDetector detector) {
        this.LastX = detector.getFocusX();
        this.LastY = detector.getFocusY();
        this.moveDisFromX = 0.0f;
        this.moveDisFromY = 0.0f;
        this.scaleStart = detector.getScaleFactor();
        if (this.mDetectorListener == null) {
            return true;
        }
        this.mDetectorListener.onScaleBegin(detector);
        return true;
    }

    public void onScaleEnd(ScaleGestureDetector detector) {
        if (this.mDetectorListener != null) {
            this.mDetectorListener.onScaleEnd(detector, this.moveDisFromX, this.moveDisFromY);
        }
    }

    public void setStartScale(float scaleFactor) {
        this.scaleStart = scaleFactor;
    }

    private float getCurrentScale(float scaleFactor) {
        this.mCurrentScale = (1.0f / this.scaleStart) * scaleFactor;
        return this.mCurrentScale;
    }

    public boolean onTouchEvent(MotionEvent ev) {
        return this.mDetector.onTouchEvent(ev);
    }
}

File: app/src/main/java/com/synapse/social/studioasinc/widget/ZoomImageViewLib/gestures/ScaleAndMoveDetector.kt
package com.synapse.social.studioasinc.widget.ZoomImageViewLib.gestures

import android.content.Context
import android.view.MotionEvent
import android.view.ScaleGestureDetector

class ScaleAndMoveDetector(context: Context, private val mDetectorListener: OnScaleAndMoveGestureListener?) : ScaleGestureDetector.OnScaleGestureListener {
    private var lastX = 0.0f
    private var lastY = 0.0f
    private var mCurrentScale = 1.0f
    private val mDetector: ScaleGestureDetector = ScaleGestureDetector(context, this)
    private var moveDisFromX = 0.0f
    private var moveDisFromY = 0.0f
    private var scaleStart = 1.0f

    override fun onScale(detector: ScaleGestureDetector): Boolean {
        val tempx = detector.focusX
        val tempy = detector.focusY
        moveDisFromX = tempx - lastX
        moveDisFromY = tempy - lastY

        mDetectorListener?.onScaleAndMove(detector, getCurrentScale(detector.scaleFactor), moveDisFromX, moveDisFromY)
        return false
    }

    override fun onScaleBegin(detector: ScaleGestureDetector): Boolean {
        lastX = detector.focusX
        lastY = detector.focusY
        moveDisFromX = 0.0f
        moveDisFromY = 0.0f
        scaleStart = detector.scaleFactor

        mDetectorListener?.onScaleBegin(detector)
        return true
    }

    override fun onScaleEnd(detector: ScaleGestureDetector) {
        mDetectorListener?.onScaleEnd(detector, moveDisFromX, moveDisFromY)
    }

    fun setStartScale(scaleFactor: Float) {
        scaleStart = scaleFactor
    }

    private fun getCurrentScale(scaleFactor: Float): Float {
        mCurrentScale = (1.0f / scaleStart) * scaleFactor
        return mCurrentScale
    }

    fun onTouchEvent(ev: MotionEvent): Boolean {
        return mDetector.onTouchEvent(ev)
    }
}
